nixos/modules/services/matrix/draupnir.nix

```diff
@@ -0,0 +1,205 @@
+              [
+                "https://matrix.to/#/#yourroom:example.org"
+                "https://matrix.to/#/#anotherroom:example.org"
+              ]
+            '';
+            description = lib.mdDoc ''
+              A list of rooms to protect (matrix.to URLs).
+              These can also be configured interactively.
+            '';
+          };
+          homeserverUrl = mkOption {
+            type = types.str;
+            default = "https://matrix.org";
+            description = lib.mdDoc ''
+              Base URL of the Matrix homeserver, that provides the Client-Server API.
+
+              If `pantalaimon.enable` is `true`, this option will become the homeserver to which `pantalaimon` connects.
+              The listen address of `pantalaimon` will then become the `homeserverUrl` of `draupnir`.
+            '';
+          };
+        };
+      };
+    };
+  };
+
+  config = mkIf config.services.draupnir.enable {
+    assertions = [
+      {
+        assertion = !(cfg.pantalaimon.enable && cfg.pantalaimon.passwordFile == null);
+        message = "Specify pantalaimon.passwordFile";
+      }
+      {
+        assertion = !(cfg.pantalaimon.enable && cfg.accessTokenFile != null);
+        message = "Do not specify accessTokenFile when using pantalaimon";
+      }
+      {
+        assertion = !(!cfg.pantalaimon.enable && cfg.accessTokenFile == null);
+        message = "Specify accessTokenFile when not using pantalaimon";
+      }
+    ];
+
+    services.pantalaimon-headless.instances."draupnir" = mkIf cfg.pantalaimon.enable
+      {
+        homeserver = cfg.homeserverUrl;
+      } // cfg.pantalaimon.options;
+
+    systemd.services.draupnir = {
+      description = "Draupnir - a moderation tool for Matrix";
+      requires = optionals (cfg.pantalaimon.enable) [
+        "pantalaimon-draupnir.service"
+      ];
+      wants = [
+        "network-online.target"
+      ];
+      after = [
+        "network-online.target"
+        "matrix-synapse.service"
+        "conduit.service"
+      ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ExecStart = "${lib.getExe pkgs.draupnir} --draupnir-config ${config}";
+        ExecStartPre = [ generateConfig ];
+        WorkingDirectory = "/var/lib/draupnir";
+        StateDirectory = "draupnir";
+        StateDirectoryMode = "0700";
+        ProtectSystem = "strict";
+        ProtectHome = true;
+        PrivateTmp = true;
+        NoNewPrivileges = true;
+        PrivateDevices = true;
+        Restart = "on-failure";
+
+        /* TODO: wait for #102397 to be resolved. Then load secrets from $CREDENTIALS_DIRECTORY+"/NAME" */
+        DynamicUser = true;
+        LoadCredential =
+          optionals (cfg.accessTokenFile != null) [
+            "access_token:${cfg.accessTokenFile}"
+          ] ++
+          optionals (cfg.pantalaimon.passwordFile != null) [
+            "pantalaimon_password:${cfg.pantalaimon.passwordFile}"
+          ];
```

Comment:
AFAIK it's because the module itself registers a pantalaimon instance? Am I missing something about how this all works here?

Reply:
I think you do. Because you are configuring another module as part of this modules config. The other module gets the passwordFile and needs to take care of how it accesses it by itself.

Reply:
Right. Makes sense, I'll remove it from the list then.